clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/MoveExecutor.scala

```diff
@@ -0,0 +1,212 @@
+      existingMetadata.pathsToMove.mkString("'", "', '", "'")
+
+    val googleCopies = existingMetadata.pathsToMove.map { p =>
+      Future(copyGoogleObject(p, destination, ioUtil))
+    }
+
+    // VERY IMPORTANT that this is lazy; we shouldn't try to delete
+    // until Clio has been updated successfully.
+    lazy val googleDeletes = existingMetadata.pathsToMove.map { p =>
+      Future(deleteGoogleObject(p, ioUtil))
+    }
+
+    for {
+      _ <- Future
+        .sequence(googleCopies)
+        .logErrorMsg(
+          s"""An error occurred while copying files in the cloud. Clio hasn't been updated,
+             |but some files for the $name may exist in two locations. Check the logs to see
+             |which copy commands failed, and why, then try re-running this command. If this
+             |can't be done, please contact the Green Team at ${ClioClientConfig.greenTeamEmail}.""".stripMargin
+        )
+      upsertResponse <- updateAndUpsert(client, existingMetadata)
+        .logErrorMsg(
+          s"""An error occurred while updating the $name record in Clio. All files associated
+             |with $name exist at both the old and new locations, but Clio only knows about the
+             |old location. Try removing the files at $destination and re-running this command.
+             |If this can't be done, please contact the Green Team at ${ClioClientConfig.greenTeamEmail}""".stripMargin
+        )
+      _ <- Future
+        .sequence(googleDeletes)
+        .logErrorMsg(
+          s"""The old files associated with the $name was not able to be deleted.
+           |Please manually delete the old files for $name at: $sourcesAsString.
+           |If this can't be done, please contact the Green Team at ${ClioClientConfig.greenTeamEmail}""".stripMargin
+        )
+    } yield {
+      logger.info(s"Successfully moved $sourcesAsString to '$destination'")
+      upsertResponse
+    }
+  }
+
+  private def copyGoogleObject(source: String,
+                               destination: String,
+                               ioUtil: IoUtil): Future[Unit] = {
+    ioUtil.copyGoogleObject(source, destination) match {
+      case 0 => Future.successful(())
+      case _ =>
+        Future.failed(
+          new Exception(
+            s"Copy files in the cloud failed from '$source' to '$destination'"
+          )
+        )
+    }
+  }
+
+  private def deleteGoogleObject(path: String, ioUtil: IoUtil): Future[Unit] = {
+    ioUtil.deleteGoogleObject(path) match {
+      case 0 => Future.successful(())
+      case _ =>
+        Future.failed(
+          new Exception(s"Deleting file in the cloud failed for path '$path'")
+        )
+    }
+  }
+
+  private def updateAndUpsert(client: ClioWebClient,
+                              existingMetadata: moveCommand.index.MetadataType)(
+    implicit credentials: HttpCredentials,
+    ec: ExecutionContext
+  ): Future[HttpResponse] = {
+    val newMetadata = if (destination.endsWith("/")) {
```

Comment:
We test `endsWith("/")` in a few places, this should be a function, e.g. `isDirectory()` on some utility class.